File: backend/matches/rating_engine.py
"""
rating_engine.py
Automated, position-specific player match rating engine based on eFootball PES 2021 individual statistics.
Calculates realistic match ratings clamped between 3.0 and 10.0 with 1 decimal precision.
"""
from decimal import Decimal


def calculate_player_rating(position, stats, match_context=None):
    """
    Calculates post-match performance rating (3.0 - 10.0) based on detailed PES individual statistics.

    :param position: Player primary position (e.g., 'CF', 'AMF', 'CB', 'GK')
    :param stats: dict containing individual PES stats
    :param match_context: optional dict (e.g. clean_sheet, team_won, yellow_cards, red_cards)
    :return: dict with 'rating' (float), 'rating_decimal' (Decimal), and 'breakdown' (list of factors)
    """
    if not stats or not isinstance(stats, dict):
        return {
            'rating': 6.0,
            'rating_decimal': Decimal('6.0'),
            'breakdown': [{'label': 'نمره پایه استاندارد', 'impact': '+6.0'}]
        }

    match_context = match_context or {}
    pos = str(position or 'CMF').upper().strip()
    breakdown = []

    # 1. Base rating
    rating = 6.0
    breakdown.append({'label': 'نمره پایه ورود به زمین', 'impact': '+6.0'})

    # Extract all metrics safely
    goals = int(stats.get('goals') or 0)
    penalty_goals = int(stats.get('penalty_goals') or 0)
    freekick_goals = int(stats.get('freekick_goals') or 0)
    assists = int(stats.get('assists') or 0)

    shots_total = int(stats.get('shots_total') or 0)
    shots_on_target = int(stats.get('shots_on_target') or 0)
    off_target_shots = max(0, shots_total - shots_on_target)

    passes_total = int(stats.get('passes_total') or 0)
    passes_completed = int(stats.get('passes_completed') or 0)
    crosses = int(stats.get('crosses') or 0)

    fouls = int(stats.get('fouls') or 0)
    offsides = int(stats.get('offsides') or 0)
    freekicks_won = int(stats.get('freekicks_won') or 0)
    corners = int(stats.get('corners') or 0)
    blocks = int(stats.get('blocks') or 0)
    minutes_played = int(stats.get('minutes_played') or 90)
    touches = int(stats.get('touches') or 0)
    dribble_distance = float(stats.get('dribble_distance') or 0.0)

    duels_total = int(stats.get('duels_total') or 0)
    duels_won = int(stats.get('duels_won') or 0)
    clearances = int(stats.get('clearances') or 0)

    # Goalkeeper metrics
    gk_shots_faced = int(stats.get('gk_shots_faced') or 0)
    gk_shots_on_target = int(stats.get('gk_shots_on_target') or 0)
    gk_saves = int(stats.get('gk_saves') or 0)
    goals_conceded = int(stats.get('goals_conceded') or 0)
    penalty_saves = int(stats.get('penalty_saves') or 0)

    # Match context
    clean_sheet = bool(match_context.get('clean_sheet', False))
    yellow_cards = int(match_context.get('yellow_cards') or stats.get('yellow_cards') or 0)
    red_cards = int(match_context.get('red_cards') or stats.get('red_cards') or 0)
    team_won = bool(match_context.get('team_won', False))

    # --- POSITION CATEGORIZATION ---
    is_goalkeeper = pos == 'GK'
    is_forward = pos in ['CF', 'SS', 'LWF', 'RWF']
    is_midfielder = pos in ['AMF', 'CMF', 'DMF', 'LMF', 'RMF']
    is_defender = pos in ['CB', 'LB', 'RB']

    # --- GOALKEEPER CALCULATIONS ---
    if is_goalkeeper:
        if gk_saves > 0:
            saves_bonus = round(gk_saves * 0.5, 2)
            rating += saves_bonus
            breakdown.append({'label': f'مهار شوت‌های درون چارچوب ({gk_saves} مهار)', 'impact': f'+{saves_bonus}'})

        if penalty_saves > 0:
            pk_save_bonus = round(penalty_saves * 1.5, 2)
            rating += pk_save_bonus
            breakdown.append({'label': f'مهار پنالتی حریف ({penalty_saves} مهار)', 'impact': f'+{pk_save_bonus}'})

        if gk_shots_on_target > 0:
            save_rate = gk_saves / gk_shots_on_target
            if save_rate >= 0.8:
                rating += 0.6
                breakdown.append({'label': f'درصد مهار عالی بالای ۸۰٪ ({int(save_rate*100)}%)', 'impact': '+0.6'})
            elif save_rate <= 0.4 and goals_conceded > 1:
                rating -= 0.5
                breakdown.append({'label': 'درصد مهار پایین شوت‌ها', 'impact': '-0.5'})

        if clean_sheet or goals_conceded == 0:
            rating += 0.8
            breakdown.append({'label': 'کلین‌شیت دروازه‌بان', 'impact': '+0.8'})
        elif goals_conceded > 0:
            conceded_penalty = round(goals_conceded * 0.35, 2)
            rating -= conceded_penalty
            breakdown.append({'label': f'گل خورده تیم ({goals_conceded} گل)', 'impact': f'-{conceded_penalty}'})

    # --- FORWARD CALCULATIONS ---
    elif is_forward:
        if goals > 0:
            g_bonus = round(goals * 1.0, 2)
            rating += g_bonus
            breakdown.append({'label': f'گل‌های زده در جریان بازی ({goals} گل)', 'impact': f'+{g_bonus}'})

        if freekick_goals > 0:
            fk_bonus = round(freekick_goals * 1.2, 2)
            rating += fk_bonus
            breakdown.append({'label': f'گل دیدنی از ضربه آزاد ({freekick_goals} گل)', 'impact': f'+{fk_bonus}'})

        if penalty_goals > 0:
            pkg_bonus = round(penalty_goals * 0.7, 2)
            rating += pkg_bonus
            breakdown.append({'label': f'گل پنالتی ({penalty_goals} گل)', 'impact': f'+{pkg_bonus}'})

        if assists > 0:
            a_bonus = round(assists * 0.7, 2)
            rating += a_bonus
            breakdown.append({'label': f'پاس گل تعیین‌کننده ({assists} پاس گل)', 'impact': f'+{a_bonus}'})

        if shots_on_target > 0:
            sot_bonus = round(shots_on_target * 0.2, 2)
            rating += sot_bonus
            breakdown.append({'label': f'شوت در چارچوب حریف ({shots_on_target} شوت)', 'impact': f'+{sot_bonus}'})

        if off_target_shots >= 3:
            rating -= 0.3
            breakdown.append({'label': f'شوت‌های ناموفق خارج چارچوب ({off_target_shots} شوت)', 'impact': '-0.3'})

        if dribble_distance >= 20.0:
            dribble_bonus = round(min(0.6, (dribble_distance / 30.0) * 0.3), 2)
            rating += dribble_bonus
            breakdown.append({'label': f'مسافت موثر دریبل ({dribble_distance:.1f} متر)', 'impact': f'+{dribble_bonus}'})

        if duels_won > 0:
            duel_bonus = round(min(0.5, duels_won * 0.15), 2)
            rating += duel_bonus
            breakdown.append({'label': f'نبردهای پیروز مهاجم ({duels_won} نبرد)', 'impact': f'+{duel_bonus}'})

        if offsides >= 2:
            off_penalty = round(min(0.4, offsides * 0.15), 2)
            rating -= off_penalty
            breakdown.append({'label': f'تله‌های آفساید حریف ({offsides} بار)', 'impact': f'-{off_penalty}'})

    # --- MIDFIELDER CALCULATIONS ---
    elif is_midfielder:
        if passes_total >= 5:
            pass_acc = passes_completed / passes_total
            if pass_acc >= 0.88:
                rating += 0.7
                breakdown.append({'label': f'دقت پاس عالی هافبک ({int(pass_acc*100)}%)', 'impact': '+0.7'})
            elif pass_acc >= 0.78:
                rating += 0.35
                breakdown.append({'label': f'دقت پاس استاندارد ({int(pass_acc*100)}%)', 'impact': '+0.35'})
            elif pass_acc <= 0.60:
                rating -= 0.4
                breakdown.append({'label': f'دقت پاس ضعیف ({int(pass_acc*100)}%)', 'impact': '-0.4'})

        if assists > 0:
            a_bonus = round(assists * 0.85, 2)
            rating += a_bonus
            breakdown.append({'label': f'پاس گل و گلسازی ({assists} پاس گل)', 'impact': f'+{a_bonus}'})

        if goals > 0:
            g_bonus = round(goals * 0.95, 2)
            rating += g_bonus
            breakdown.append({'label': f'گلزنی هافبک ({goals} گل)', 'impact': f'+{g_bonus}'})

        if blocks > 0:
            block_bonus = round(min(0.6, blocks * 0.3), 2)
            rating += block_bonus
            breakdown.append({'label': f'سد توپ و قطع مسیر پاس ({blocks} بار)', 'impact': f'+{block_bonus}'})

        if clearances > 0:
            clear_bonus = round(min(0.5, clearances * 0.25), 2)
            rating += clear_bonus
            breakdown.append({'label': f'دفع و بازپس‌گیری توپ ({clearances} بار)', 'impact': f'+{clear_bonus}'})

        if duels_won > 0:
            duel_bonus = round(min(0.6, duels_won * 0.2), 2)
            rating += duel_bonus
            breakdown.append({'label': f'پیروزی در دوئل‌های میانه میدان ({duels_won} نبرد)', 'impact': f'+{duel_bonus}'})

        if crosses >= 3:
            rating += 0.2
            breakdown.append({'label': f'ارسال‌های موفق سانتر ({crosses} سانتر)', 'impact': '+0.2'})

    # --- DEFENDER CALCULATIONS ---
    elif is_defender:
        if clearances > 0:
            clear_bonus = round(min(0.9, clearances * 0.35), 2)
            rating += clear_bonus
            breakdown.append({'label': f'دفع و پوشش مدافع ({clearances} دفع توپ)', 'impact': f'+{clear_bonus}'})

        if blocks > 0:
            block_bonus = round(min(0.8, blocks * 0.4), 2)
            rating += block_bonus
            breakdown.append({'label': f'سد توپ و بلاک شوت ({blocks} بار)', 'impact': f'+{block_bonus}'})

        if duels_won > 0:
            duel_bonus = round(min(0.6, duels_won * 0.2), 2)
            rating += duel_bonus
            breakdown.append({'label': f'پیروزی در نبردهای تن‌به‌تن ({duels_won} نبرد)', 'impact': f'+{duel_bonus}'})

        if duels_total > duels_won and (duels_total - duels_won) >= 3:
            rating -= 0.3
            breakdown.append({'label': 'شکست در نبردهای دفاعی', 'impact': '-0.3'})

        if clean_sheet or goals_conceded == 0:
            rating += 0.7
            breakdown.append({'label': 'کلین‌شیت خط دفاع', 'impact': '+0.7'})
        elif goals_conceded > 1:
            conceded_penalty = round(min(0.8, (goals_conceded - 1) * 0.25), 2)
            rating -= conceded_penalty
            breakdown.append({'label': f'آسیب‌پذیری خط دفاع ({goals_conceded} گل خورده)', 'impact': f'-{conceded_penalty}'})

        if goals > 0:
            g_bonus = round(goals * 1.1, 2)
            rating += g_bonus
            breakdown.append({'label': f'گلزنی درخشان مدافع ({goals} گل)', 'impact': f'+{g_bonus}'})

    # --- UNIVERSAL DISCIPLINARY & MATCH IMPACTS ---
    if fouls >= 3:
        foul_penalty = round(min(0.5, fouls * 0.1), 2)
        rating -= foul_penalty
        breakdown.append({'label': f'خطاهای مکرر بازیکن ({fouls} خطا)', 'impact': f'-{foul_penalty}'})

    if freekicks_won >= 2:
        rating += 0.2
        breakdown.append({'label': f'گرفتن ضربه آزاد خطرناک ({freekicks_won} ضربه)', 'impact': '+0.2'})

    if yellow_cards > 0:
        yc_penalty = round(yellow_cards * 0.5, 2)
        rating -= yc_penalty
        breakdown.append({'label': f'کارت زرد انضباطی ({yellow_cards} کارت)', 'impact': f'-{yc_penalty}'})

    if red_cards > 0:
        rating -= 2.0
        breakdown.append({'label': 'اخراج با کارت قرمز مستقیم', 'impact': '-2.0'})

    if team_won:
        rating += 0.2
        breakdown.append({'label': 'پیروزی نهایی تیم', 'impact': '+0.2'})

    # Minute scaling: if played less than 20 minutes, scale deviation towards 6.0
    if minutes_played < 25:
        deviation = rating - 6.0
        rating = 6.0 + (deviation * (minutes_played / 30.0))

    # Clamp rating to standard football match scale [3.0, 10.0]
    final_rating = max(3.0, min(10.0, round(rating, 1)))

    return {
        'rating': float(final_rating),
        'rating_decimal': Decimal(str(final_rating)),
        'breakdown': breakdown
    }

File: frontend/src/utils/ratingEngine.js
/**
 * ratingEngine.js
 * Client-side calculation engine mirroring backend/matches/rating_engine.py.
 * Calculates player match performance ratings (3.0 - 10.0) based on eFootball PES 2021 stats.
 */

export function calculatePlayerRating(position, stats = {}, matchContext = {}) {
  if (!stats || typeof stats !== 'object') {
    return {
      rating: 6.0,
      breakdown: [{ label: 'نمره پایه ورود به زمین', impact: '+6.0' }]
    };
  }

  const pos = String(position || 'CMF').toUpperCase().trim();
  const breakdown = [];
  let rating = 6.0;
  breakdown.push({ label: 'نمره پایه ورود به زمین', impact: '+6.0' });

  // Numerical stat extractions
  const goals = Number(stats.goals) || 0;
  const penaltyGoals = Number(stats.penalty_goals) || 0;
  const freekickGoals = Number(stats.freekick_goals) || 0;
  const assists = Number(stats.assists) || 0;

  const shotsTotal = Number(stats.shots_total) || 0;
  const shotsOnTarget = Number(stats.shots_on_target) || 0;
  const offTargetShots = Math.max(0, shotsTotal - shotsOnTarget);

  const passesTotal = Number(stats.passes_total) || 0;
  const passesCompleted = Number(stats.passes_completed) || 0;
  const crosses = Number(stats.crosses) || 0;

  const fouls = Number(stats.fouls) || 0;
  const offsides = Number(stats.offsides) || 0;
  const freekicksWon = Number(stats.freekicks_won) || 0;
  const blocks = Number(stats.blocks) || 0;
  const minutesPlayed = Number(stats.minutes_played) || 90;
  const dribbleDistance = Number(stats.dribble_distance) || 0;

  const duelsTotal = Number(stats.duels_total) || 0;
  const duelsWon = Number(stats.duels_won) || 0;
  const clearances = Number(stats.clearances) || 0;

  // Goalkeeper metrics
  const gkShotsOnTarget = Number(stats.gk_shots_on_target) || 0;
  const gkSaves = Number(stats.gk_saves) || 0;
  const penaltySaves = Number(stats.penalty_saves) || 0;

  // Match context
  const cleanSheet = Boolean(matchContext.clean_sheet);
  const goalsConceded = Number(matchContext.goals_conceded ?? stats.goals_conceded ?? 0);
  const yellowCards = Number(matchContext.yellow_cards ?? stats.yellow_cards ?? 0);
  const redCards = Number(matchContext.red_cards ?? stats.red_cards ?? 0);
  const teamWon = Boolean(matchContext.team_won);

  const isGoalkeeper = pos === 'GK';
  const isForward = ['CF', 'SS', 'LWF', 'RWF'].includes(pos);
  const isMidfielder = ['AMF', 'CMF', 'DMF', 'LMF', 'RMF'].includes(pos);
  const isDefender = ['CB', 'LB', 'RB'].includes(pos);

  // --- GOALKEEPER ---
  if (isGoalkeeper) {
    if (gkSaves > 0) {
      const bonus = Number((gkSaves * 0.5).toFixed(2));
      rating += bonus;
      breakdown.push({ label: `مهار شوت‌های درون چارچوب (${gkSaves} مهار)`, impact: `+${bonus}` });
    }

    if (penaltySaves > 0) {
      const bonus = Number((penaltySaves * 1.5).toFixed(2));
      rating += bonus;
      breakdown.push({ label: `مهار پنالتی حریف (${penaltySaves} مهار)`, impact: `+${bonus}` });
    }

    if (gkShotsOnTarget > 0) {
      const saveRate = gkSaves / gkShotsOnTarget;
      if (saveRate >= 0.8) {
        rating += 0.6;
        breakdown.push({ label: `درصد مهار عالی بالای ۸۰٪ (${Math.round(saveRate * 100)}%)`, impact: '+0.6' });
      } else if (saveRate <= 0.4 && goalsConceded > 1) {
        rating -= 0.5;
        breakdown.push({ label: 'درصد مهار پایین شوت‌ها', impact: '-0.5' });
      }
    }

    if (cleanSheet || goalsConceded === 0) {
      rating += 0.8;
      breakdown.push({ label: 'کلین‌شیت دروازه‌بان', impact: '+0.8' });
    } else if (goalsConceded > 0) {
      const pen = Number((goalsConceded * 0.35).toFixed(2));
      rating -= pen;
      breakdown.push({ label: `گل خورده تیم (${goalsConceded} گل)`, impact: `-${pen}` });
    }
  }

  // --- FORWARD ---
  else if (isForward) {
    if (goals > 0) {
      const bonus = Number((goals * 1.0).toFixed(2));
      rating += bonus;
      breakdown.push({ label: `گل‌های زده در جریان بازی (${goals} گل)`, impact: `+${bonus}` });
    }

    if (freekickGoals > 0) {
      const bonus = Number((freekickGoals * 1.2).toFixed(2));
      rating += bonus;
      breakdown.push({ label: `گل دیدنی ضربه آزاد (${freekickGoals} گل)`, impact: `+${bonus}` });
    }

    if (penaltyGoals > 0) {
      const bonus = Number((penaltyGoals * 0.7).toFixed(2));
      rating += bonus;
      breakdown.push({ label: `گل پنالتی (${penaltyGoals} گل)`, impact: `+${bonus}` });
    }

    if (assists > 0) {
      const bonus = Number((assists * 0.7).toFixed(2));
      rating += bonus;
      breakdown.push({ label: `پاس گل تعیین‌کننده (${assists} پاس گل)`, impact: `+${bonus}` });
    }

    if (shotsOnTarget > 0) {
      const bonus = Number((shotsOnTarget * 0.2).toFixed(2));
      rating += bonus;
      breakdown.push({ label: `شوت در چارچوب حریف (${shotsOnTarget} شوت)`, impact: `+${bonus}` });
    }

    if (offTargetShots >= 3) {
      rating -= 0.3;
      breakdown.push({ label: `شوت‌های ناموفق خارج چارچوب (${offTargetShots} شوت)`, impact: '-0.3' });
    }

    if (dribbleDistance >= 20.0) {
      const bonus = Number(Math.min(0.6, (dribbleDistance / 30.0) * 0.3).toFixed(2));
      rating += bonus;
      breakdown.push({ label: `مسافت موثر دریبل (${dribbleDistance.toFixed(1)} متر)`, impact: `+${bonus}` });
    }

    if (duelsWon > 0) {
      const bonus = Number(Math.min(0.5, duelsWon * 0.15).toFixed(2));
      rating += bonus;
      breakdown.push({ label: `نبردهای پیروز مهاجم (${duelsWon} نبرد)`, impact: `+${bonus}` });
    }

    if (offsides >= 2) {
      const pen = Number(Math.min(0.4, offsides * 0.15).toFixed(2));
      rating -= pen;
      breakdown.push({ label: `تله آفساید حریف (${offsides} بار)`, impact: `-${pen}` });
    }
  }

  // --- MIDFIELDER ---
  else if (isMidfielder) {
    if (passesTotal >= 5) {
      const passAcc = passesCompleted / passesTotal;
      if (passAcc >= 0.88) {
        rating += 0.7;
        breakdown.push({ label: `دقت پاس عالی هافبک (${Math.round(passAcc * 100)}%)`, impact: '+0.7' });
      } else if (passAcc >= 0.78) {
        rating += 0.35;
        breakdown.push({ label: `دقت پاس استاندارد (${Math.round(passAcc * 100)}%)`, impact: '+0.35' });
      } else if (passAcc <= 0.60) {
        rating -= 0.4;
        breakdown.push({ label: `دقت پاس ضعیف (${Math.round(passAcc * 100)}%)`, impact: '-0.4' });
      }
    }

    if (assists > 0) {
      const bonus = Number((assists * 0.85).toFixed(2));
      rating += bonus;
      breakdown.push({ label: `پاس گل و گلسازی (${assists} پاس گل)`, impact: `+${bonus}` });
    }

    if (goals > 0) {
      const bonus = Number((goals * 0.95).toFixed(2));
      rating += bonus;
      breakdown.push({ label: `گلزنی هافبک (${goals} گل)`, impact: `+${bonus}` });
    }

    if (blocks > 0) {
      const bonus = Number(Math.min(0.6, blocks * 0.3).toFixed(2));
      rating += bonus;
      breakdown.push({ label: `سد توپ و قطع مسیر پاس (${blocks} بار)`, impact: `+${bonus}` });
    }

    if (clearances > 0) {
      const bonus = Number(Math.min(0.5, clearances * 0.25).toFixed(2));
      rating += bonus;
      breakdown.push({ label: `دفع و بازپس‌گیری توپ (${clearances} بار)`, impact: `+${bonus}` });
    }

    if (duelsWon > 0) {
      const bonus = Number(Math.min(0.6, duelsWon * 0.2).toFixed(2));
      rating += bonus;
      breakdown.push({ label: `پیروزی در نبردهای میانه میدان (${duelsWon} نبرد)`, impact: `+${bonus}` });
    }

    if (crosses >= 3) {
      rating += 0.2;
      breakdown.push({ label: `ارسال‌های موفق سانتر (${crosses} سانتر)`, impact: '+0.2' });
    }
  }

  // --- DEFENDER ---
  else if (isDefender) {
    if (clearances > 0) {
      const bonus = Number(Math.min(0.9, clearances * 0.35).toFixed(2));
      rating += bonus;
      breakdown.push({ label: `دفع و پوشش مدافع (${clearances} دفع توپ)`, impact: `+${bonus}` });
    }

    if (blocks > 0) {
      const bonus = Number(Math.min(0.8, blocks * 0.4).toFixed(2));
      rating += bonus;
      breakdown.push({ label: `سد توپ و بلاک شوت (${blocks} بار)`, impact: `+${bonus}` });
    }

    if (duelsWon > 0) {
      const bonus = Number(Math.min(0.6, duelsWon * 0.2).toFixed(2));
      rating += bonus;
      breakdown.push({ label: `پیروزی در نبردهای تن‌به‌تن (${duelsWon} نبرد)`, impact: `+${bonus}` });
    }

    if (duelsTotal > duelsWon && (duelsTotal - duelsWon) >= 3) {
      rating -= 0.3;
      breakdown.push({ label: 'شکست در نبردهای دفاعی', impact: '-0.3' });
    }

    if (cleanSheet || goalsConceded === 0) {
      rating += 0.7;
      breakdown.push({ label: 'کلین‌شیت خط دفاع', impact: '+0.7' });
    } else if (goalsConceded > 1) {
      const pen = Number(Math.min(0.8, (goalsConceded - 1) * 0.25).toFixed(2));
      rating -= pen;
      breakdown.push({ label: `آسیب‌پذیری خط دفاع (${goalsConceded} گل خورده)`, impact: `-${pen}` });
    }

    if (goals > 0) {
      const bonus = Number((goals * 1.1).toFixed(2));
      rating += bonus;
      breakdown.push({ label: `گلزنی درخشان مدافع (${goals} گل)`, impact: `+${bonus}` });
    }
  }

  // --- DISCIPLINARY & MATCH OUTCOME ---
  if (fouls >= 3) {
    const pen = Number(Math.min(0.5, fouls * 0.1).toFixed(2));
    rating -= pen;
    breakdown.push({ label: `خطاهای مکرر (${fouls} خطا)`, impact: `-${pen}` });
  }

  if (freekicksWon >= 2) {
    rating += 0.2;
    breakdown.push({ label: `گرفتن ضربه آزاد خطرناک (${freekicksWon} ضربه)`, impact: '+0.2' });
  }

  if (yellowCards > 0) {
    const pen = Number((yellowCards * 0.5).toFixed(2));
    rating -= pen;
    breakdown.push({ label: `کارت زرد انضباطی (${yellowCards} کارت)`, impact: `-${pen}` });
  }

  if (redCards > 0) {
    rating -= 2.0;
    breakdown.push({ label: 'اخراج با کارت قرمز مستقیم', impact: '-2.0' });
  }

  if (teamWon) {
    rating += 0.2;
    breakdown.push({ label: 'پیروزی نهایی تیم', impact: '+0.2' });
  }

  // Minute scaling
  if (minutesPlayed < 25) {
    const deviation = rating - 6.0;
    rating = 6.0 + (deviation * (minutesPlayed / 30.0));
  }

  const finalRating = Math.max(3.0, Math.min(10.0, Math.round(rating * 10) / 10));

  return {
    rating: Number(finalRating.toFixed(1)),
    breakdown
  };
}
